ykrt/src/compile/jitc_yk/jit_ir/mod.rs

```diff
@@ -1,9 +1,45 @@
 //! The JIT's Intermediate Representation (IR).
 //!
-//! This is the IR created by [trace_builder] and which is then optimised. The IR can feel a little
-//! odd at first because we store indexes into vectors rather than use direct references. This
-//! allows us to squeeze the amount of the memory used down (and also bypasses issues with
-//! representing graph structures in Rust, but that's slightly accidental).
+//! This is the IR created by [trace_builder] and which is then analysed and optimised. The IR has
+//! the following properties:
+//!
+//!   1. We use vectors rather than direct references. This allows us to squeeze the amount of the
+//!      memory used down (and also bypasses issues with representing graph structures in Rust, but
+//!      that's slightly accidental).
+//!
+//!   2. The IR is designed to allow most manipulations to be performed without requiring
+//!      reallocation.
+//!
+//! This second property has some surprising consequences:
+//!
+//!   1. We have to introduce `Proxy*` instructions which semi-transparently "forward" an
+//!      instruction's results on. For example if we IR in memory stored along these lines:
+//!
+//!        ```text
+//!        %0: ...
+//!        %1: 1i8
+//!        %2: add %0, %1
+//!        ```
+//!
+//!      Then instruction `%1` is a `ProxyConst`. We very rarely want to view the IR as it is
+//!      directly stored in memory. Instead, we nearly always want to pretend that the IR really
+//!      looks like:
+//!
+//!        ```text
+//!        %0: ...
+//!        %2: add %0, 1i8
+//!        ```
+//!
+//!      i.e. we want to pretend that the `%1` doesn't exist and that the reference to `%1` is
+//!      actually just a constant.
+//!
+//!   2. We never store [Operand]s directly: we only ever store [PackedOperand]s. Whenever we want
+//!      to get an [Operand] we *must* use [PackedOperand::unpack]. `unpack` does the necessary
+//!      deproxying.
+//!
+//!   3. In nearly all cases we should view the IR as deproxied: unless you explicitly intend to
```

Comment:
If we really never want this, why don't we make `Module::inst` the no proxy version and `Module::inst_with_proxies` the other?

Reply:
I'm pretty sure we can come up with better names but I'd rather get a bit more experience with the API before worrying *too* much about it.